Initial attempt at supporting ESP-IDF 5.0.0

kbx81 opened the pull request:
# What does this implement/fix?

This is just a batch of changes I found necessary to get a successful build with IDF 5.0.

It's...a mess...but we have to start somewhere. My hope is that others will find at least some of this work helpful as we move towards full, proper support of the updated framework.

In general, this is very much a WIP and I don't expect we'll merge this specific PR...but, who knows. 😄

Version 6.0 of the platform also seems to complain about `printf` (and siblings) all over the place. I'm guessing this is some compiler setting but I don't know how and didn't take the time to look up how to fix that...so for now I have a commit with that nonsense. You'll see it complain in any other component(s) where those functions are used.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

Note that, for now, this **requires** importing the mDNS component via @jesserockz's #4000 PR; but, with this, I was able to get a successful build. That said, mDNS **will not actually work** because we have more work to do to get the hostname set correctly. You'll probably notice warnings about this if/when you build.

That said, _if you run this on your ESP, you can still do updates OTA, but you'll need to use the device's IP as it won't respond to any hostname._

```yaml
# Example config.yaml
esphome:
  platformio_options:
    build_flags:
      - -DCONFIG_FREERTOS_ENABLE_BACKWARD_COMPATIBILITY

esp32:
  board: featheresp32
  framework:
    type: esp-idf
    version: "5.0.0"
    platform_version: "6.0.0"
    components:
      - name: esp-protocols
        source: github://espressif/esp-protocols

mdns:
  disabled: true
```

**NOTE:** I also ran into an issue building as described in [this thread](https://github.com/espressif/esp-idf/issues/8489); read the comments for the resolution (basically just run `install.sh` followed by `. export.sh`).

## Checklist:
  - [x] The code change is tested and works locally...mostly, kind of.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

kbx81 committed:
wifi: fix build with ESP-IDF >= 5

ESP-IDF 4.1 introduced the esp-netif API as successor to the tcp_adapter
API. The tcp_adapter API was removed in ESP-IDF 5.0.0. Part of the wifi
component was already migrated to the new API. Migrate the leftover uses
of the old API to the new API to fix build on ESP-IDF >= 5.

The version of ESP-IDF currently in use (4.4.4) supports the new API, so
we don't need any if guards to maintain backwards compatibility.

Also replace xQueueHandle, which is a pre FreeRTOS v8.0.0 data type,
with QueueHandle_t, so we don't need to enable backward compatibility
(CONFIG_FREERTOS_ENABLE_BACKWARD_COMPATIBILITY).

This reverts part of commit d42f35de5d54 to wifi_component_esp_idf.cpp,
as the esp-netif API handles that internally.

[replace pre FreeRTOS v8.0.0 data type, add commit message]
Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

kbx81 committed:
ota: fix build with ESP-IDF >= 5

As of version 5, esp_task_wdt_init() takes a struct as argument. We also
need to include spi_flash_mmap.h.

[split unrelated change into separate commits, maintain ESP-IDF < 5
compat, use esp_task_wdt_reconfigure, add commit message]
Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

kbx81 committed:
core: fix build with ESP-IDF >= 5

These header files already existed in ESP-IDF < 5 so skip if guards.

[add commit message]
Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

oxan commented:
Some discussion on the discord: https://discord.com/channels/429907082951524364/1076282397903048725

stintel committed:
docker: install python3-venv

As of version 6.0.1, platform-espressif32 requires python3-venv.
Switching between esp-idf 4.4.4 and 5.0 causes problems with esp-idf
python dependencies installed by PlatformIO. They've solved this by
using venv. Install python3-venv so that platform-espressif32 6.0.1 and
later can be used, and we don't need to wipe the dependencies manually
when switching esp-idf versions.

Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

stintel committed:
esp32: fix build with ESP-IDF >= 5

We need to include esp_timer.h to be able to use esp_timer_get_time().
This header existed in ESP-IDF < 5 so we don't need if guards.

Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

stintel committed:
mdns: fix build with ESP-IDF >= 5

In ESP-IDF 5.0.0, the mdns component was removed and moved to another
repository. Since the mdns component in esphome is always built, we
need to add the mdns component from the esp-protocols repository. This
component depends on ESP-IDF >= 5.0, so we need to add a version guard.

Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

stintel committed:
requirements: add pyparsing >= 3.0

ESP-IDF >= 5.0 requires pyparsing's rest_of_file, which was introduced
in version 3.0.

Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

stintel commented:
Tested the changes I just pushed on ESP32-C3-DevKitM-1 and WEMOS LOLIN D32 PRO using the following version combinations:

* esp-idf 4.4.4 + platform-espressif 5.3.0
* esp-idf 5.0.1 + platform-espressif 6.0.0

oxan commented on esphome/components/mdns/__init__.py:
```diff
@@ -79,6 +84,14 @@ async def to_code(config):
         elif CORE.is_rp2040:
             cg.add_library("LEAmDNS", None)
 
+    if CORE.using_esp_idf and (framework_ver.major, framework_ver.minor) >= (5, 0):
```
Minor nit; but you can compare `cv.Version` instances directly:
```suggestion
    if CORE.using_esp_idf and framework_ver >= cv.Version(5, 0, 0):
```
(or maybe even drop the `framework_ver` variable).

oxan commented on esphome/core/config.py:
```diff
@@ -383,6 +386,9 @@ async def to_code(config):
     if CORE.using_arduino:
         CORE.add_job(add_arduino_global_workaround)
 
+    if CORE.using_esp_idf and (framework_ver.major, framework_ver.minor) >= (5, 0):
```
same here.

oxan reviewed (approved):
LGTM.

stintel commented:
Tested the changes on ESP32-C3-DevKitM-1 and WEMOS LOLIN D32 PRO using the following version combinations:

    esp-idf 4.4.4 + platform-espressif 5.3.0
    esp-idf 5.0.1 + platform-espressif 6.0.0

This time also with IPv6 enabled and disabled, due to the changes I made to resolve conflicts after d42f35de5d54.
@jesserockz can you please verify I've correctly resolved these conflicts?

I also noticed that switching esp-idf versions requires wiping esp-idf's python dependencies installed by platformio. This is exactly what they fixed in https://github.com/platformio/platform-espressif32/commit/8507dd8e7fc6d3fb56ec5372fc24a3044226a058, so I've added a commit to install python3-venv.

Afterwards retested with platform-espressif32 6.0.1 and 6.1.0.

jesserockz reviewed (approved):
Changes look fine. Will give them a quick test locally during the week if not today.

jesserockz commented on esphome/core/config.py:
```diff
@@ -383,6 +384,11 @@ async def to_code(config):
+    if CORE.using_esp_idf and CORE.data[KEY_CORE][KEY_FRAMEWORK_VERSION] >= cv.Version(
+        5, 0, 0
+    ):
+        cg.add_build_flag("-Wno-format")
```
Is this still required after https://github.com/esphome/esphome/pull/4671 ?

stintel replied:
No. Dropped the commit.

stintel commented:
Anything still blocking this PR from being merged?
